octavia_f5/controller/worker/f5agent_driver.py

```diff
@@ -81,38 +81,34 @@ def tenant_update(bigip,
             continue
 
         # Create generic application
-        app = Application(constants.APPLICATION_GENERIC,
-                          label=loadbalancer.id)
+        app = Application(constants.APPLICATION_GENERIC, label=loadbalancer.id)
 
-        # attach listeners with ESDs / L7Policies
+        # Attach L7Policies and ESDs to listeners
         for listener in loadbalancer.listeners:
             if utils.pending_delete(listener):
                 continue
 
-            profiles = {}
+            # Translate L7policies to endpoint policies
             for l7policy in listener.l7policies:
                 if utils.pending_delete(l7policy):
                     continue
+                app.add_endpoint_policy(
+                    m_policy.get_name(l7policy.id),
+                    m_policy.get_endpoint_policy(l7policy)
+                )
```

Comment:
I would like to move the endpoint policy addition to the service.py, else in case of l7policies that are ESD, they are appended but unused in the application.

Reply:
Since endpoint policies are not added to the service but the application, I'd rather keep the code at this place and instead check for and skip ESD-policies.
I'll add this fix in another commit, please comment on that one if you have suggestions for further improvements.

Reply:
Still you need to iterate twice over l7policies, one time outside of the service and one time inside the service. Since endpoint policies are always referencend by services, the generation of the l7policies and their references is better done inside the service and then return as another entity.

Reply:
Very good point, thanks. I pushed a fix.